src/features/user/components/CouchDB/CouchDB.tsx

```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { Divider, Form, Header, Input } from 'semantic-ui-react';
+import { Link, Redirect, RouteProps } from 'react-router-dom';
+import { saveUsername, savePassword, saveURL } from '../../state/ui/CouchDB.action';
+import { finishCouchDBSetting } from 'features/user/state/ui/CouchDB.action'
+import './index.css';
+
+export interface CouchDBPropsT {
+  isSignedIn: boolean;
+  username: string;
+  password: string;
+  url: string;
+  saveUsername: typeof saveURL;
+  savePassword: typeof saveURL;
+  saveURL: typeof saveURL;
+  finishCouchDBSetting: typeof finishCouchDBSetting;
+};
+
+class CouchDB extends React.Component<CouchDBPropsT & RouteProps, {}> {
+  componentDidMount() {
+  }
+
+  render() {
+    if (this.props.isSignedIn) return <Redirect to="/" />;
+
+    return (
+      <div className="container-raised-desktop">
+        <Header as="h2" icon="id card outline" content="CouchDB setup" />
+        <Divider />
+        <p>Setup CouchDB.</p>
+        <Form onSubmit={this.props.finishCouchDBSetting}>
+          <div className="couchdb--input-field">
+            <Input
+              fluid required icon="hand point right outline" iconPosition="left"
+              placeholder="CouchDB Username"
+              name="username" type="text"
+              value={this.props.username}
+              onChange={(_, { value }) => this.props.saveUsername(value)}
+            />
+            <Input
+              fluid required icon="hand point right outline" iconPosition="left"
+              placeholder="CouchDB Password"
+              name="username" type="password"
+              value={this.props.password}
+              onChange={(_, { value }) => this.props.savePassword(value)}
+            />
+            <Input
+              fluid required icon="hand point right outline" iconPosition="left"
+              placeholder="CouchDB URL"
+              name="url" type="text"
+              value={this.props.url}
+              onChange={(_, { value }) => this.props.saveURL(value)}
+            />
+            <button className="ui primary button" type="submit">Submit</button>
```

Comment:
In the [SignIn](https://github.com/ayastreb/money-tracker/blob/master/src/features/user/components/SignIn/SignIn.tsx#L129) component we disable input fields and set primary button `loading` to `true` once the form is submitted - this gives user a feedback that something is loading and they should wait.

It seems like in CouchDB component we don't handle loading state, so when user clicks "Submit" - the app will start loading the data in background, but UI won't update until all the data is synced, which might take a while.

Do you think you could add loading state handling in this component?

Reply:
I'll have a try tomorrow.

Reply:
As I'm new to web development, it's not clear to me what's the best way to accomplish this.

Reading code in `SignIn` component,

- it uses `props.location.hash` to determine whether render a loading page
- `props.sendCodeStatus` determines whether button is disabled.

For `CouchDB` settings component, I can think of the following solution:

- Add a boolean state `data_syncing` to `CouchDBStateT`
- Use this state to change `CouchDB` component behavior

Is there any better way to do this?

Reply:
I see there's a `isRunning` state in [sync.js](https://github.com/ayastreb/money-tracker/blob/master/src/reducers/ui/sync.js#L23). But I can't figure out how to access this state in CouchDB container.

Reply:
To be honest it's been a while since I worked on this code myself, so I don't remember how certain things work :)
General idea is to add `isRunning` field in CouchDB state with `false` value by defaul and change the value to `true` when `finishCouchDBSetting` action is called. You can then read this field in container with `mapStateToProps`.

We don't have to make this change in this MR, it can be a separate improvement later, if you don't have time for it now

Reply:
Thanks for the suggestion. I will try to understand state management more before I continue on this improvement.